File: base/src/main/java/com/union/bangbang/base/base/BaseActivity.java
package com.union.bangbang.base.base;


import android.content.Context;
import android.os.Bundle;
import androidx.annotation.IdRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

import dagger.android.AndroidInjection;
import dagger.android.support.DaggerAppCompatActivity;

/**
 * @name toto
 * @anthor bangbang QQ:740090077
 * @time 2019-02-05 16:17
 * 只有编译器可能不骗你。
 */

public abstract class BaseActivity<Binding extends ViewDataBinding> extends DaggerAppCompatActivity {
    protected Binding binding;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        initView();
    }

    private void initView() {
        binding = DataBindingUtil.setContentView(this, getLayoutId());
    }
    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
    protected abstract  int getLayoutId();

    public Binding getBinding() {
        if(binding==null)initView();
        return binding;
    }
}

File: todokotlin/src/main/java/com/union/bangbang/todokotlin/base/BaseActivity.kt
package com.union.bangbang.todokotlin.base

import android.arch.lifecycle.AndroidViewModel
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.union.bangbang.todokotlin.BR
import dagger.android.AndroidInjection

/**
 * @name toto
 * @anthor bangbang QQ:740090077
 * @time 2018/12/16 3:45 PM
 * 只有编译器可能不骗你。
 */
abstract class BaseActivity<Binding:ViewDataBinding> : AppCompatActivity() {
    lateinit var binding: Binding
    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        initViewDataBinding()
    }

    companion object Route {
        fun onStartActivity(context: Context, clazz: Class<*>) {
            val intent = Intent(context, clazz)
            context.startActivity(intent)
        }
    }
    fun initViewDataBinding(){
        binding = DataBindingUtil.setContentView(this, getLayoutId())
        binding.setVariable(BR.vm,initViewModel())
    }
    abstract fun getLayoutId():Int
    abstract fun initViewModel():AndroidViewModel
}